iOS improvements

kambala-decapitator opened the pull request:
## Description
I want to make 32+64 bit device+simulator build of FFmpeg using modern toolchain.

## Type of Change
- New feature
- Bug fix

## Checks
- [ ] Changes support all platforms (`Android`, `iOS`, `macOS`, `tvOS`)
- [ ] Breaks existing functionality
- [x] Implementation is completed, not half-done 
- [ ] Is there another PR already created for this feature/bug fix

## Tests
Made 32+64 bit device+simulator (without Catalyst) build of FFmpeg using Xcode 12.5.1.

```
./ios.sh --disable-arm64-mac-catalyst --disable-x86-64-mac-catalyst --speed --lts --enable-ios-audiotoolbox --enable-ios-avfoundation --enable-ios-bzip2 --enable-ios-zlib --enable-ios-libiconv --enable-ios-videotoolbox --no-framework

Building ffmpeg-kit LTS shared library for iOS

Architectures: armv7, armv7s, arm64, arm64e, i386, x86-64, arm64-simulator
Libraries: ios-zlib, ios-audiotoolbox, ios-bzip2, ios-videotoolbox, ios-avfoundation, ios-libiconv

Downloading sources: ok

Building armv7 platform targeting iOS SDK 9.3 and Mac Catalyst 13.0


ffmpeg: ok

ffmpeg-kit: ok

Building armv7s platform targeting iOS SDK 9.3 and Mac Catalyst 13.0


ffmpeg: ok

ffmpeg-kit: ok

Building arm64 platform targeting iOS SDK 9.3 and Mac Catalyst 13.0


ffmpeg: ok

ffmpeg-kit: ok

Building arm64e platform targeting iOS SDK 9.3 and Mac Catalyst 13.0


ffmpeg: ok

ffmpeg-kit: ok

Building i386 platform targeting iOS SDK 9.3 and Mac Catalyst 13.0


ffmpeg: ok

ffmpeg-kit: ok

Building x86-64 platform targeting iOS SDK 9.3 and Mac Catalyst 13.0


ffmpeg: ok

ffmpeg-kit: ok

Building arm64-simulator platform targeting iOS SDK 9.3 and Mac Catalyst 13.0


ffmpeg: ok

ffmpeg-kit: ok
```

tanersener commented:
Thanks. 

Have you tested your changes on the `development` branch? Because the thread local issue on `i386` is already fixed there. See the [updated cflags](https://github.com/tanersener/ffmpeg-kit/blob/development/scripts/function-ios.sh#L146).

Regarding the VideoToolbox patch, if `FFmpeg` has already updated the source code of `libavutil/hwcontext_videotoolbox.c`, wouldn't be better to use the new `FFmpeg` version instead of patching something that will be replaced soon. As you know, we are updating the `FFmpeg` version before the releases and will do that for the next release as well.

kambala-decapitator commented:
> Thanks.
> 
> Have you tested your changes on the `development` branch? Because the thread local issue on `i386` is already fixed there. See the [updated cflags](https://github.com/tanersener/ffmpeg-kit/blob/development/scripts/function-ios.sh#L146).

just built all architectures and confirm that my i386 commit is no longer needed, dropped it.

Also enabled building arm64e regardless of the iOS deployment target.

> Regarding the VideoToolbox patch, if `FFmpeg` has already updated the source code of `libavutil/hwcontext_videotoolbox.c`, wouldn't be better to use the new `FFmpeg` version instead of patching something that will be replaced soon.

after double-checking the FFmpeg's commit I think that they actually did a wrong thing. The best would probably be submitting a patch to FFmpeg to bring back the availability check and detect if `__builtin_available` symbol is available or not, but I don't really have a desire to deal with that. For this project a simple patch would do.

Should I drop my current VTB commit? I can also open another PR as soon as FFmpeg version is bumped to apply a patch instead of simple `sed` commands.

> As you know, we are updating the `FFmpeg` version before the releases and will do that for the next release as well.

this I didn't know, sorry. I found this project only as a convenient way to build "minimal" FFmpeg for iOS.

tanersener commented:
> Also enabled building arm64e regardless of the iOS deployment target.

Sorry, what's the problem about `arm64e` or other 32-bit architectures? The idea behind `disable_ios_architecture_not_supported_on_detected_sdk_version` method is to automatically disable not supported architectures. It was working fine the last time I checked.

> after double-checking the FFmpeg's commit I think that they actually did a wrong thing.

Okay, I don't know the details but using `if/else` macros instead of `__builtin_available` doesn't look wrong to me. What is the issue in the updated `libavutil/hwcontext_videotoolbox.c`?

kambala-decapitator commented:
> > Also enabled building arm64e regardless of the iOS deployment target.
> 
> Sorry, what's the problem about `arm64e` or other 32-bit architectures? The idea behind `disable_ios_architecture_not_supported_on_detected_sdk_version` method is to automatically disable not supported architectures. It was working fine the last time I checked.

Started typing big comment, went to check something in the script and found out that you simply had an error in arm64e detection :) Dropped previous commit and pushed a fix.

> > after double-checking the FFmpeg's commit I think that they actually did a wrong thing.
> 
> Okay, I don't know the details but using `if/else` macros instead of `__builtin_available` doesn't look wrong to me. What is the issue exactly in the updated `libavutil/hwcontext_videotoolbox.c`?

`__builtin_available` allows weak linking to symbols, but FFmpeg decided to actually drop the right way and make simple headers check (because someone tested with a very old toolchain / SDK). Consider example: you build using macOS SDK 10.13 and set deployment target to macOS 10.12. With the new way the macro resolves to 1 and new symbol is used in the code. In the best case you receive compile error, in the worst - runtime crash on macOS 10.12 because such symbol doesn't exist there. But with weak linking you'd have simply used the fallback symbol (if any).

tanersener commented:
So, you suggest using detected sdk version to disable `arm64e`. Okay, makes sense. What about 32-bit architectures, the check about them is also dropped?

Regarding `videotoolbox` patch, this is how one of the symbols used in the `if/else` macros is defined in the `SDK`. Numbers are different than your example. I assume you say the current macro based check will cause a runtime crash if macOS SDK is 10.12 but deployment target is 10.10, right?

```CV_EXPORT const CFStringRef CV_NONNULL kCVImageBufferYCbCrMatrix_ITU_R_2020 __OSX_AVAILABLE_STARTING(__MAC_10_11,__IPHONE_9_0);```

kambala-decapitator commented:
> What about 32-bit architectures, the check about them is also dropped?

I've double-checked my change and found out that it's no longer needed actually, thanks for pointing on that! In the `main` branch there was a detection error (mistake opposite to the arm64e case :) ):

```sh
> ag --literal disable_ios_architecture_not_supported_on_detected_sdk_version
ios.sh
163:disable_ios_architecture_not_supported_on_detected_sdk_version "${ARCH_ARMV7}" "${DETECTED_IOS_SDK_VERSION}"
164:disable_ios_architecture_not_supported_on_detected_sdk_version "${ARCH_ARMV7S}" "${DETECTED_IOS_SDK_VERSION}"
165:disable_ios_architecture_not_supported_on_detected_sdk_version "${ARCH_I386}" "${DETECTED_IOS_SDK_VERSION}"
```

but in the dev branch it's correct: you check deployment target. So I've dropped my commit.

FYI when trying to build for 32-bit with a higher deployment target compile error appears:

> clang: error: invalid iOS deployment version '-target armv7-apple-ios11.0', iOS 10 is the maximum deployment target for 32-bit targets [-Winvalid-ios-deployment-target]

<hr/>

> I assume you say the current macro based check will cause a runtime crash if macOS SDK is 10.12 but deployment target is 10.10, right?

yes, you got it right. Attempt to access `kCVImageBufferYCbCrMatrix_ITU_R_2020 ` on 10.10 (or iOS < 9) would result in a crash.

BTW I always wondered: why do you make LTS build with Xcode 7? That's such an "ancient" toolchain...

tanersener commented:
> yes, you got it right. Attempt to access kCVImageBufferYCbCrMatrix_ITU_R_2020 on 10.10 (or iOS < 9) would result in a crash.

I tested this by modifying the source code and setting deployment target to `macOS 10.10.3`. And, compilation failed. It seems to me like there is no risk of a runtime crash for these symbols. `clang` will see that those symbols are not available on that target and break the build. As it happened in my test.

```
libavutil/hwcontext_videotoolbox.c:427:23: error: 'kCVImageBufferYCbCrMatrix_ITU_R_2020' is only available on macOS 10.11 or newer [-Werror,-Wunguarded-availability]
        colormatrix = kCVImageBufferYCbCrMatrix_ITU_R_2020;
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode-v12.5.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/System/Library/Frameworks/CoreVideo.framework/Headers/CVImageBuffer.h:76:40: note: 'kCVImageBufferYCbCrMatrix_ITU_R_2020' has been marked as being introduced in macOS 10.11 here, but the deployment target is macOS 10.10.3
CV_EXPORT const CFStringRef CV_NONNULL kCVImageBufferYCbCrMatrix_ITU_R_2020 __OSX_AVAILABLE_STARTING(__MAC_10_11,__IPHONE_9_0);         // CFString
                                       ^
libavutil/hwcontext_videotoolbox.c:427:23: note: enclose 'kCVImageBufferYCbCrMatrix_ITU_R_2020' in a __builtin_available check to silence this warning
        colormatrix = kCVImageBufferYCbCrMatrix_ITU_R_2020;
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [libavutil/hwcontext_videotoolbox.o] Error 1
```

> BTW I always wondered: why do you make LTS build with Xcode 7? That's such an "ancient" toolchain...

It was the oldest release I managed to download and test with the library. There is no other reason.

kambala-decapitator commented:
> And, compilation failed.

yes, fortunately `-Werror` is used in FFmpeg build, otherwise it would've been just a warning.

But I hope you understand that this way a valid use case is broken where weak linking is the exact solution.

tanersener commented:
Patching a source file, build file, etc. is something we try to avoid generally. Because maintaining them is not easy. But we still use them as the last resort if there is no other solution.

Here, in this case we are talking about a patch but I don't see something failing. `-Werror` or another flag. It doesn't matter what it is. Original source code of `ffmpeg` doesn't fail. And, we shouldn't spend our time and effort to patch something that doesn't fail. Maybe it is not perfect but it still works and is more maintainable than a patch.

kambala-decapitator commented:
tried to build FFmpeg for macOS manually and understood that the error appears at configure time.

Regarding `kCVImageBufferYCbCrMatrix_ITU_R_2020` and similar symbols, FFmpeg uses raw string fallback which I guess would be fine at runtime.

I also noticed that `clock_gettime` isn't used when deployment target is low for the same reason. From development for Apple platforms POV FFmpeg is doing it wrong by not using weak linking IMO.

So, I should drop my VTB commit since you're going to update to the latest FFmpeg commit, shouldn't I?

tanersener commented:
I already updated the `ffmpeg` version on `development` a few days ago in [this commit](https://github.com/tanersener/ffmpeg-kit/commit/eee591774605fd289e7766c46c1209fb83dd46e3). You can get the updates and see if something still needs to be patched.

You also have a commit about `allow building any architecture when --no-framework is specified`. What was the use case there?

kambala-decapitator commented:
> I already updated the `ffmpeg` version on `development` a few days ago in [this commit](https://github.com/tanersener/ffmpeg-kit/commit/eee591774605fd289e7766c46c1209fb83dd46e3). You can get the updates and see if something still needs to be patched.

yes, now my edits don't apply as expected, dropped the commit.

> You also have a commit about `allow building any architecture when --no-framework is specified`. What was the use case there?

You have conditions to disable some archs due to impossibility to put them in a lipo'd binary. But with `--no-framework` lipo doesn't take place hence no need to disable anything.

tanersener commented:
> You have conditions to disable some archs due to impossibility to put them in a lipo'd binary. But with --no-framework lipo doesn't take place hence no need to disable anything.

Yeah, I'm aware of that. I was trying to understand why you do that. The scenario, where you need to build all architectures but don't need to package them. Anyway, thanks for the PR and the changes.
